


<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Welcome to RetSynth’s documentation! &#8212; RetSynth 1 documentation</title>
    <link rel="stylesheet" href="_static/p_sphinx_theme.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/local_fonts.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/jquery.cookie.js"></script>
    <script src="_static/p_sphinx_theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
            <meta name="viewport" content="width=device-width, initial-scale=1">
  </head><body>
      <div class="relbar-top">
            
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
      <li><a href="#">RetSynth 1 documentation</a> &#187;</li>

        <li class="nav-item nav-item-this"><a href="">Welcome to RetSynth’s documentation!</a></li> 
      </ul>
    </div>
      </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-retsynth-s-documentation">
<h1>Welcome to RetSynth’s documentation!<a class="headerlink" href="#welcome-to-retsynth-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>RetSynth is a tool, developed with python, which identfies enzyme/reaction pairs that are required in a user specified microbial organisms to produce a target chemical compound.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<dl>
<dt>RetSynth works on operating systems:</dt><dd><p>Windows</p>
<p>Mac</p>
<p>Linux</p>
<p>Cygwin</p>
</dd>
<dt>Required Non-python Dependencies:</dt><dd><p><a class="reference external" href="http://ftp.gnu.org/gnu/glpk/">GLPK</a> (v4.64 preferred)</p>
<p><a class="reference external" href="http://www.graphviz.org/">GraphViz</a> (only if <code class="docutils literal notranslate"><span class="pre">--figures_graphviz</span></code> option is being used, See Visualization_graphviz option below)</p>
</dd>
<dt>Python Dependencies:</dt><dd><p><a class="reference external" href="https://pypi.org/project/PuLP/">PULP</a></p>
<p><a class="reference external" href="https://pypi.org/project/glpk/">glpk</a></p>
<p><a class="reference external" href="http://cobrapy.readthedocs.io/en/latest/">CobraPy</a></p>
<p><a class="reference external" href="https://pypi.python.org/pypi/python-libsbml-experimental/5.10.0">libsbml</a></p>
<p><a class="reference external" href="https://pypi.python.org/pypi/pygraphviz">PyGraphViz</a> (only if <code class="docutils literal notranslate"><span class="pre">--figures_graphviz</span></code> option is being used,  See Visualization_graphviz option below)</p>
<p><a class="reference external" href="https://pypi.org/project/bs4/">bs4</a></p>
<p><a class="reference external" href="https://pypi.org/project/pubchempy/">pubchempy</a></p>
<p><a class="reference external" href="https://pypi.org/project/openpyxl/">openpyxl</a></p>
<p><a class="reference external" href="https://pypi.org/project/scipy/">scipy</a></p>
<p><a class="reference external" href="https://pypi.org/project/tqdm/">tqdm</a></p>
<p><a class="reference external" href="https://pypi.org/project/soupsieve/">soupsieve</a></p>
<p><a class="reference external" href="https://pypi.org/project/filelock/">filelock</a></p>
<p><a class="reference external" href="https://pypi.org/project/distlib/">distlib</a></p>
<p><a class="reference external" href="https://pypi.org/project/jsonschema/">jsonschema</a></p>
<p><a class="reference external" href="https://pypi.org/project/matplotlib/">matplotlib</a></p>
</dd>
</dl>
</div>
</div>
<div class="section" id="purpose">
<h1>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h1>
<p>In the process of bioengineering a microbial organism, identifying reactions/enzymes to transform into an organism for optimal production of a target compound is extremely difficult due to the vast number of reactions/enzymes in numerous organisms that maybe suitable for optimal production.</p>
<p>By assembling a database of genome-wide metabolic networks for a plethora of microbial organisms and framing this information into an integer linear program, we can identify the optimal number of enzyme/reaction pairs that need to be added into an organism for optimal synthesis of a compound.  Specifically the goal is to identify the minimal number of reactions/enzymes as this is beneficial for metabolic engineering because less genetic manipulation would be required.  Using flux balance analysis FBA RetSynth can predict production yields of a compound in a select microbial organism.</p>
<p>The overarching goal of RetSynth is to streamline an arduous and complex step of bioengineering a microbial organism, which will enable scientists to inexpensively expedite the production of important target compounds.</p>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p>git clone <a class="reference external" href="https://github.com/sandialabs/RetSynth.git">https://github.com/sandialabs/RetSynth.git</a></p></li>
<li><p>Enter the master directory and run <code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">setup.py</span> <span class="pre">install</span></code></p></li>
</ol>
</div>
</div>
<div class="section" id="retsynth-workflow">
<h1>RetSynth Workflow<a class="headerlink" href="#retsynth-workflow" title="Permalink to this headline">¶</a></h1>
<div class="section" id="construction-of-metabolic-database">
<h2>1. Construction of Metabolic Database<a class="headerlink" href="#construction-of-metabolic-database" title="Permalink to this headline">¶</a></h2>
<p>In order for RetSynth to identify potential reactions/enzymes to engineer to an organism for synthesis of a target compound a database of genome-wide metabolic information for numerous microbial organisms must be available to the software.  RetSynth builds an <a class="reference external" href="https://docs.python.org/2/library/sqlite3.html/">SQLite</a> database.
RetSynth can compile a metabolic database which can include metabolic information from the following Repositories:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://www.patricbrc.org/">PATRIC</a></p></li>
<li><p><a class="reference external" href="https://metacyc.org/">MetaCyc</a></p></li>
<li><p><a class="reference external" href="http://www.genome.jp/kegg/kegg1.html">KEGG</a></p></li>
<li><p><a class="reference external" href="http://minedatabase.mcs.anl.gov/#/home">Metabolic In Silico Network Expansion Database</a></p></li>
<li><p><a class="reference external" href="http://lcsb-databases.epfl.ch/atlas/">ATLAS of BioChemistry</a></p></li>
<li><p><a class="reference external" href="https://www.spresi.com/">SPRESI</a></p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The SPRESI database has to be purchased by the user and for RetSynth to integrate SPRESI data into its database it currently has to be in the <a class="reference external" href="http://www.daylight.com/meetings/mug05/Kappler/ctfile.pdf">Rxnfiles</a> (.rdf) format.</p>
</div>
<p>To build a database the user must identify the metabolic repository with which the database is to be constructed (<code class="docutils literal notranslate"><span class="pre">--patric</span></code>, <code class="docutils literal notranslate"><span class="pre">--metacyc</span></code>, <code class="docutils literal notranslate"><span class="pre">--kegg</span></code>, <code class="docutils literal notranslate"><span class="pre">--atlas</span></code>, <code class="docutils literal notranslate"><span class="pre">--mine</span></code> and/or <code class="docutils literal notranslate"><span class="pre">--SPRESI</span></code>). Furthermore the directory or file which contains the raw metabolic repository (specifically needed for –kbase, –metacyc, –mine and –atlas) information must be specified:</p>
<p><code class="docutils literal notranslate"><span class="pre">-a_dir</span> <span class="pre">or</span> <span class="pre">--atlas_dump_directory</span></code> for atlas data (csv files) with the <code class="docutils literal notranslate"><span class="pre">--atlas</span></code> option,</p>
<p><code class="docutils literal notranslate"><span class="pre">-m_dir</span> <span class="pre">or</span> <span class="pre">--mine_dump_directory</span></code> for mine data (msp files) with the <code class="docutils literal notranslate"><span class="pre">--mine</span></code> option,</p>
<p><code class="docutils literal notranslate"><span class="pre">s_dir</span> <span class="pre">or</span> <span class="pre">--spresi_dump_directory</span></code> for spresi data (rdf files) with the <code class="docutils literal notranslate"><span class="pre">--SPRESI</span></code> option and</p>
<p><code class="docutils literal notranslate"><span class="pre">-mc</span> <span class="pre">or</span> <span class="pre">--metacyc_addition</span></code> (which is the xml (metabolic-reactions.xml) file for the metacyc reactions repository) with the –metacyc option.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">--kegg</span></code> option does not require a directory specification as RetSynth connects directly to the kegg database.</p>
<p>Additionally the <code class="docutils literal notranslate"><span class="pre">--patric_models</span></code> option connects directly to the patric server using the <a class="reference external" href="https://dx.doi.org/doi:10.1093/bioinformatics/btx185">mackinac</a> python library which was integrated into RetSynth. In order to use patric the user must input a patric username <code class="docutils literal notranslate"><span class="pre">--patric_username</span></code> and patric password <code class="docutils literal notranslate"><span class="pre">--patric_password</span></code>.  Note the user can select which patric models are integrated into the RetSynth database (defualt is currently Escherichia coli DH1, Streptomyces venezuelae ATCC 10712 and Corynebacterium argentoratense DSM 44202) by utilizing the <code class="docutils literal notranslate"><span class="pre">--patricfile</span></code>.
The <code class="docutils literal notranslate"><span class="pre">--patricfile</span></code> option is a csv file which tells which RetSynth which genomes from patric to download into the database. In order to generate your own list of genomes:</p>
<ol class="arabic simple">
<li><p>go to <a class="reference external" href="https://www.patricbrc.org/">PATRIC</a> and click on <code class="docutils literal notranslate"><span class="pre">organisms</span></code> then <code class="docutils literal notranslate"><span class="pre">All</span> <span class="pre">Bacteria</span></code>.  Click on the <code class="docutils literal notranslate"><span class="pre">Genomes</span></code> tab and select the genomes you want to include in the database.</p></li>
</ol>
<img alt="_images/PATRICweb1.png" src="_images/PATRICweb1.png" />
<ol class="arabic simple" start="2">
<li><p>Click the download button selecting the csv option.</p></li>
</ol>
<img alt="_images/PATRICweb2.png" src="_images/PATRICweb2.png" />
<p>If the user is constructing a database using multiple repositories it is suggested that the <code class="docutils literal notranslate"><span class="pre">--inchidb</span></code> option is specified although this option drastically increases the time required to build the database.  This converts all compound IDs to their inchi values and therefore elimiates duplicate entries of metabolites and reactions from the different metabolic repositories.  If this option is not selected KEGG compound IDs will be used. Additionally, the user must designate a database name with the option <code class="docutils literal notranslate"><span class="pre">-gdb</span> <span class="pre">or</span> <span class="pre">--generate_database</span></code>. In the event that a database has already been built and the user wants to use that database instead of building a new one, the database can be accessed by utilizing the option <code class="docutils literal notranslate"><span class="pre">-db</span> <span class="pre">or</span> <span class="pre">--database</span></code> with the name of the database.</p>
</div>
<div class="section" id="generating-stoichiometric-matrix-for-reactions-and-compounds-in-the-metabolic-database">
<h2>2. Generating Stoichiometric Matrix for Reactions and Compounds in the Metabolic Database<a class="headerlink" href="#generating-stoichiometric-matrix-for-reactions-and-compounds-in-the-metabolic-database" title="Permalink to this headline">¶</a></h2>
<p>To find the minimal reactions needed to optimally produce a compound, a stoichiometric matrix for all compounds and reactions in a database is required.  The stoichiometric matrix is a mathmatical representation of the reactions and the compounds they degrade and synthesize.  If a stoichiometric has not been generated for a database the user must specify <code class="docutils literal notranslate"><span class="pre">-gdbc</span> <span class="pre">or</span> <span class="pre">--generate_database_constraints</span></code> with the file name which will generate .constraints file.</p>
<p>Because the .constraints file does not have to be recreated if the user wants to perform multiple searchs for multiple target compounds within the same database, the user can specify the option <code class="docutils literal notranslate"><span class="pre">-dbc</span> <span class="pre">or</span> <span class="pre">--database_constraints</span></code> with the .constraints file name. This option saves time as initial construction of constraints file is time-consuming.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Each constructed database requires its own .constraints file.  Additionally if a database has both chemical and biological reactions in it and the user only wants say biological reactions it is good to generate a new constraints file (i.e. constraintfile_bio.constraints) with the <code class="docutils literal notranslate"><span class="pre">-gdbc</span></code> option.</p>
</div>
</div>
<div class="section" id="target-compounds-input-file">
<h2>3. Target Compounds (input file)<a class="headerlink" href="#target-compounds-input-file" title="Permalink to this headline">¶</a></h2>
<dl class="simple">
<dt>RetSynth also requires a list of compounds and the organism they are desired to be synthesized in <code class="docutils literal notranslate"><span class="pre">-t</span> <span class="pre">or</span> <span class="pre">--targets</span></code>.  In the input file, which must be a tab-deliminated text file, there are several options that need to be specified:</dt><dd><ol class="arabic simple">
<li><p>The target compound ID, pubchem ID, inchi value or name (required)</p></li>
<li><p>The desired organism ID or organism name (optional, however, if not specified the software proceeds to identify the number of reactions that need to be added to each organism in the database to produce the target)</p></li>
<li><p>Reaction IDs that the user does not want the software to identify as a potential method of producing the target (optional)</p></li>
</ol>
</dd>
</dl>
<p>Example input file 1:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#compoundid   organismID    ignore reactions</span>
<span class="n">cpdT</span> <span class="n">organism_name</span>  <span class="n">rxn10</span>
</pre></div>
</div>
<p>Example input file 2:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#pubchem organism</span>
<span class="mi">263</span>   <span class="n">organism_name</span>
</pre></div>
</div>
<p>Example input file 3:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#name organismid</span>
<span class="mi">1</span><span class="o">-</span><span class="n">butanol</span>   <span class="n">organism_name1</span><span class="p">,</span> <span class="n">organism_name2</span>
</pre></div>
</div>
<p>Example input file 4:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#name organismid</span>
<span class="n">InChI</span><span class="o">=</span><span class="mi">1</span><span class="n">S</span><span class="o">/</span><span class="n">C4H10O</span><span class="o">/</span><span class="n">c1</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="mi">3</span><span class="o">-</span><span class="mi">4</span><span class="o">-</span><span class="mi">5</span><span class="o">/</span><span class="n">h5H</span><span class="p">,</span><span class="mi">2</span><span class="o">-</span><span class="mi">4</span><span class="n">H2</span><span class="p">,</span><span class="mi">1</span><span class="n">H3</span>  <span class="n">organism_name1</span><span class="p">,</span> <span class="n">organism_name2</span>
</pre></div>
</div>
</div>
<div class="section" id="examples-running-retsynth">
<h2>4. Examples running RetSynth<a class="headerlink" href="#examples-running-retsynth" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="examples-for-running-retsynth">
<h2>Examples for running RetSynth<a class="headerlink" href="#examples-for-running-retsynth" title="Permalink to this headline">¶</a></h2>
<p>Example 1: If database and constraint file have not already been generated:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">RetSynthGC</span> <span class="kn">import</span> <span class="n">RetSynthGC</span>
<span class="n">RetSynthGC</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span><span class="p">,</span> <span class="n">generate_database</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">generate_database_constraint</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span><span class="p">,</span> <span class="n">patric_models</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">patric_username</span><span class="o">=</span><span class="n">Username</span><span class="p">,</span> <span class="n">patric_password</span><span class="o">=</span><span class="n">Password</span><span class="p">,</span> <span class="n">metacyc</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metacyc_addition</span><span class="o">=./</span><span class="n">metabolic</span><span class="o">-</span><span class="n">reactions</span><span class="o">.</span><span class="n">xml</span><span class="p">)</span>
<span class="ow">or</span>
<span class="n">rs_gc</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">t</span> <span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">gdb</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">db</span> <span class="o">-</span><span class="n">gdbc</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span> <span class="o">--</span><span class="n">patric_models</span> <span class="o">--</span><span class="n">patric_username</span> <span class="n">Username</span> <span class="o">--</span><span class="n">patric_password</span> <span class="n">Password</span> <span class="o">--</span><span class="n">metacyc</span><span class="o">=</span><span class="kc">True</span> <span class="o">--</span><span class="n">metacyc_addition</span><span class="o">=./</span><span class="n">metabolic</span><span class="o">-</span><span class="n">reactions</span><span class="o">.</span><span class="n">xml</span>
</pre></div>
</div>
<p>Example 2: If database and constraint file have been generated:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">RetSynthGC</span> <span class="kn">import</span> <span class="n">RetSynthGC</span>
<span class="n">RetSynthGC</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">database_constraints</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span><span class="p">)</span>
<span class="ow">or</span>
<span class="n">rs_gc</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">t</span> <span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">db</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">db</span> <span class="o">-</span><span class="n">dbc</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span>
</pre></div>
</div>
<p>Example 3: If database has been generated but constraint file has not been generated:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">RetSynthGC</span> <span class="kn">import</span> <span class="n">RetSynthGC</span>
<span class="n">RetSynthGC</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">generate_database_constraints</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span><span class="p">)</span>
<span class="ow">or</span>
<span class="n">rs_gc</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">t</span> <span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">db</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">db</span> <span class="o">-</span><span class="n">gdbc</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span>
</pre></div>
</div>
<p>The remainder of this documentation goes over other options that can be added for further results and analysis provided by RetSynth.</p>
<div class="section" id="results">
<h3>Results<a class="headerlink" href="#results" title="Permalink to this headline">¶</a></h3>
<p>An output file is generated named optimal_pathways.txt which gives the number pathways found which could synthesize the target compound with the mimimal number of reactions.</p>
<p>Example optimal_pathways.txt file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SHORTEST</span> <span class="n">PATHS</span> <span class="n">FOR</span> <span class="n">cpdT_c0</span> <span class="n">cpdT_c0</span> <span class="ow">in</span> <span class="n">target</span> <span class="n">organism</span> <span class="n">test1</span><span class="o">.</span><span class="n">xml</span>
<span class="n">Solution</span> <span class="mi">1</span>
<span class="n">rxn7_c0</span>  <span class="n">rxn7_c0</span>  <span class="n">forward</span>  <span class="kc">None</span>  <span class="mi">1</span> <span class="n">number</span> <span class="n">of</span> <span class="n">species</span> <span class="n">that</span> <span class="n">contain</span> <span class="n">this</span> <span class="n">reaction</span>
   <span class="n">cpdZ_c0</span>  <span class="n">cpdI_c0</span> <span class="n">reactant</span>
   <span class="n">cpdE_c0</span>  <span class="n">cpdB_c0</span> <span class="n">product</span>
<span class="n">rxn8_c0</span>  <span class="n">rxn8_c0</span>  <span class="n">forward</span>  <span class="kc">None</span>  <span class="mi">1</span> <span class="n">number</span> <span class="n">of</span> <span class="n">species</span> <span class="n">that</span> <span class="n">contain</span> <span class="n">this</span> <span class="n">reaction</span>
   <span class="n">cpdE_c0</span>  <span class="n">cpdB_c0</span> <span class="n">reactant</span>
   <span class="n">cpdX_c0</span>  <span class="n">cpdX_c0</span> <span class="n">product</span>
<span class="n">rxn11_c0</span> <span class="n">rxn11_c0</span> <span class="n">forward</span>  <span class="kc">None</span>  <span class="mi">1</span> <span class="n">number</span> <span class="n">of</span> <span class="n">species</span> <span class="n">that</span> <span class="n">contain</span> <span class="n">this</span> <span class="n">reaction</span>
   <span class="n">cpdY_c0</span>  <span class="n">cpdY_c0</span> <span class="n">reactant</span>
   <span class="n">cpdT_c0</span>  <span class="n">cpdT_c0</span> <span class="n">product</span>
<span class="n">rxn9_c0</span>  <span class="n">rxn9_c0</span>  <span class="n">forward</span>  <span class="kc">None</span>  <span class="mi">1</span> <span class="n">number</span> <span class="n">of</span> <span class="n">species</span> <span class="n">that</span> <span class="n">contain</span> <span class="n">this</span> <span class="n">reaction</span>
   <span class="n">cpdX_c0</span>  <span class="n">cpdX_c0</span> <span class="n">reactant</span>
   <span class="n">cpdW_c0</span>  <span class="n">cpdW_c0</span> <span class="n">product</span>
<span class="n">rxn10_c0</span> <span class="n">rxn10_c0</span> <span class="n">forward</span>  <span class="kc">None</span>  <span class="mi">1</span> <span class="n">number</span> <span class="n">of</span> <span class="n">species</span> <span class="n">that</span> <span class="n">contain</span> <span class="n">this</span> <span class="n">reaction</span>
   <span class="n">cpdW_c0</span>  <span class="n">cpdW_c0</span> <span class="n">reactant</span>
   <span class="n">cpdY_c0</span>  <span class="n">cpdY_c0</span> <span class="n">product</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="gene-compatability-gc">
<h2>Gene Compatability (GC)<a class="headerlink" href="#gene-compatability-gc" title="Permalink to this headline">¶</a></h2>
<p>RetSynth has a built in module which after identifying a novel path to compound production will identify the optimal gene sequences for each gene/enzyme in the novel pathway.
A host microbial organism will have a unique codon bias, meaning that the organism better translates particular codon sequences to amino acids.  By optimizing the codon sequences to
match the host organisms codon bias we can increase efficiency of gene experession of the genes/enzymes in the novel pathway and subsequently optimal production of the target compound. The
software comes with 4 organism codon usage tables which include Escherichia coli DH1, Streptomyces venezuelae ATCC 10712, Corynebacterium argentoratense DSM 44202 and Rhodospirillum rubrum.</p>
<p>The user can generate there own codon bias usage table by:
1) going to the <a class="reference external" href="hive.biochemistry.gwu.edu/review/codon">HIVE-CUTS</a> database and inputing the name of the organism they wish to create a table for in the search box and clicking the submit button.</p>
<img alt="_images/HIVE1.png" src="_images/HIVE1.png" />
<ol class="arabic simple" start="2">
<li><p>Click on codon usage bias table tab and copy and paste the table into a text file and then save that text file</p></li>
</ol>
<img alt="_images/HIVE2.png" src="_images/HIVE2.png" />
<ol class="arabic simple" start="3">
<li><p>Then specify the path to this file using the <code class="docutils literal notranslate"><span class="pre">user_cai_table</span></code> option when running the code</p></li>
</ol>
<p>To implement the gene compatability feature:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">RetSynthGC</span> <span class="kn">import</span> <span class="n">RetSynthGC</span>
<span class="n">RetSynthGC</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">database_constraints</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span><span class="p">,</span> <span class="n">gene_compatability</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cai_threshold</span><span class="o">=</span><span class="mf">0.50</span><span class="p">)</span>
<span class="ow">or</span>
<span class="n">rs_gc</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">t</span> <span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">db</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">db</span> <span class="o">-</span><span class="n">gdbc</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span> <span class="o">-</span><span class="n">gc</span> <span class="o">-</span><span class="n">cai_thresh</span> <span class="mf">0.5</span>
</pre></div>
</div>
<p>The codon adaptation index (CAI) is a measure of how close the gene sequence is to a given reference set (in our case the host organism)</p>
</div>
<div class="section" id="flux-balance-analysis-fba">
<h2>Flux Balance Analysis (FBA)<a class="headerlink" href="#flux-balance-analysis-fba" title="Permalink to this headline">¶</a></h2>
<p>RetSynth, using the FBA tool <a class="reference external" href="http://cobrapy.readthedocs.io/en/latest/">CobraPy</a>, can simulate the metabolic activity of an organism and calculated the maximum theoretical yield of a target compound using the identified enzyme/reaction pairs.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently FBA can only be run on organisms added from <a class="reference external" href="https://www.patricbrc.org/">PATRIC</a>.</p>
</div>
<p>To run FBA the <code class="docutils literal notranslate"><span class="pre">-fba</span> <span class="pre">or</span> <span class="pre">--flux_balance_analysis</span></code> option must be specified.  Using the option <code class="docutils literal notranslate"><span class="pre">-media</span> <span class="pre">or</span> <span class="pre">--media_for_FBA</span></code> will specify that the organism’s metabolism will be simulated on a user specified media.  Currently these media files can also be obtained from  <a class="reference external" href="https://kbase.us/">KBase</a>.  Finally, the user can perform knockouts, <code class="docutils literal notranslate"><span class="pre">-ko</span> <span class="pre">or</span> <span class="pre">--knockouts</span></code> of each reaction in the organism, which will help in the understanding of which enzymes, reactions and metabolic pathways are important for synthesis of the target compound.</p>
<div class="section" id="fba-results">
<h3>FBA Results<a class="headerlink" href="#fba-results" title="Permalink to this headline">¶</a></h3>
<p>Several files are generated when the <code class="docutils literal notranslate"><span class="pre">-fba</span></code> option is run. flux_output.txt is the main output file which gives the objective function values before and after the reactions identified in the previous step are added to the organism.  Additionally, reactions whose flux changes 2.5 fold from when metabolism is simulated prior to the reactions being add to after they are added are shown in this output document.</p>
<p>Example flux_output.txt:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">FBA</span> <span class="n">Solutions</span> <span class="k">for</span> <span class="n">cpdT_c0</span>
<span class="mf">0.0</span>   <span class="mf">1000.0</span> <span class="n">objective</span> <span class="n">function</span> <span class="n">solutions</span> <span class="k">for</span> <span class="n">wild</span><span class="o">-</span><span class="nb">type</span> <span class="ow">and</span> <span class="n">mutant</span><span class="p">,</span> <span class="n">respectively</span>

<span class="n">Fluxes</span> <span class="n">that</span> <span class="n">differ</span> <span class="n">by</span> <span class="mf">1.5</span> <span class="n">fold</span> <span class="k">for</span> <span class="n">reactions</span> <span class="n">between</span> <span class="n">wildtype</span> <span class="ow">and</span> <span class="n">mutant</span><span class="p">:</span>
      <span class="n">wildtype</span> <span class="n">flux</span>  <span class="n">mutantflux</span>
   <span class="n">Trans_Z_c0</span>  <span class="mf">0.0</span>   <span class="mf">1000.0</span>
   <span class="n">EX_I_e0</span>  <span class="mf">0.0</span>   <span class="o">-</span><span class="mf">1000.0</span>

<span class="n">Fluxes</span> <span class="k">for</span> <span class="n">added</span> <span class="n">reactions</span> <span class="ow">in</span> <span class="n">mutant</span><span class="p">:</span>
   <span class="n">rxn8_c0</span>  <span class="n">rxn8_c0</span>  <span class="mf">1000.0</span>
   <span class="n">rxn7_c0</span>  <span class="n">rxn7_c0</span>  <span class="mf">1000.0</span>
   <span class="n">rxn10_c0</span> <span class="n">rxn10_c0</span> <span class="mf">1000.0</span>
   <span class="n">rxn11_c0</span> <span class="n">rxn11_c0</span> <span class="mf">1000.0</span>
   <span class="n">Sink_E2</span>  <span class="kc">None</span>  <span class="mf">1000.0</span>
   <span class="n">rxn9_c0</span>  <span class="n">rxn9_c0</span>  <span class="mf">1000.0</span>

<span class="n">External</span> <span class="n">pathway</span> <span class="k">with</span> <span class="n">most</span> <span class="n">flux</span><span class="p">:</span>
   <span class="n">Path</span> <span class="mi">1</span>   <span class="p">[</span><span class="s1">&#39;rxn7_c0&#39;</span><span class="p">,</span> <span class="s1">&#39;rxn8_c0&#39;</span><span class="p">,</span> <span class="s1">&#39;rxn11_c0&#39;</span><span class="p">,</span> <span class="s1">&#39;rxn9_c0&#39;</span><span class="p">,</span> <span class="s1">&#39;rxn10_c0&#39;</span><span class="p">]</span>
   <span class="n">Total</span> <span class="n">flux</span> <span class="n">through</span> <span class="n">path</span><span class="p">:</span> <span class="mf">5000.0</span>
</pre></div>
</div>
<p>Additionaly, theoretical_yield.txt file is generated which gives the percent theoretical yields of the target compounds produced in a select organism. (Note: this is only accurate if <code class="docutils literal notranslate"><span class="pre">-fba</span></code>  is run on with <code class="docutils literal notranslate"><span class="pre">-media</span></code> option)</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">-ko</span></code> parameter is used, essentialrxns_output.txt file is generated which lists all the reactions, which when knocked out prevent production of a target compound.</p>
</div>
</div>
<div class="section" id="visualization-chemdraw">
<h2>Visualization_chemdraw<a class="headerlink" href="#visualization-chemdraw" title="Permalink to this headline">¶</a></h2>
<p>This module generates figures of the pathways needed to synthesize the target compound in chemdraw (cdxml) format.  To use this module the user must specify the optin <code class="docutils literal notranslate"><span class="pre">figures_chemdraw</span></code>.  Additionally, if chemical reactions have been added to the database and the user wants all chemical reaction information to be shown in the figure (i.e solvent, catalyst, pressure…) then the user must specify <code class="docutils literal notranslate"><span class="pre">--show_chemical_rxn_info</span></code> option.</p>
<img alt="_images/but-2-one.png" src="_images/but-2-one.png" />
<p>Gray compounds indicate the compound is native to the target organism</p>
<p>Black square nodes show reactions that need to be added to target organism</p>
<p>Blue compounds indicate that the compound is not native to the target organism</p>
<p>Red compound indicates the target compound</p>
</div>
<div class="section" id="visualization-graphviz">
<h2>Visualization_graphviz<a class="headerlink" href="#visualization-graphviz" title="Permalink to this headline">¶</a></h2>
<p>This module, which requires <a class="reference external" href="http://www.graphviz.org/">GraphViz</a> , generates graphs of the reactions that are required to synthesize the compound in a desired organism.  To use this module the user must specify the option <code class="docutils literal notranslate"><span class="pre">--figures_graphviz</span></code> .  Additionaly if the user wants the chemical structures to be used in the figures instead of round nodes option <code class="docutils literal notranslate"><span class="pre">--images</span></code> must be specified.
Example Figure:</p>
<img alt="_images/SC_graph_1_propanol_Escherichia_coli_DH1.xml.png" src="_images/SC_graph_1_propanol_Escherichia_coli_DH1.xml.png" />
<dl>
<dt>This figure shows the reactions and compounds needed to produce a target compound (1-propanol) in a target organism (E. Coli).</dt><dd><p>Gray compounds indicate the compound is native to the target organism</p>
<p>Black square nodes show reactions that need to be added to target organism</p>
<p>Blue compounds indicate that the compound is not native to the target organism</p>
<p>Red compound indicates the target compound</p>
</dd>
</dl>
<p>In this case there are a multiple pathways (with the same minimal number of reactions) to get to 1-propanol in E. Coli.  The red edges indicate that those are the reactions that produce the maximum theoretical yield of the target compound.</p>
</div>
<div class="section" id="visualize-results-in-html">
<h2>Visualize results in html<a class="headerlink" href="#visualize-results-in-html" title="Permalink to this headline">¶</a></h2>
<p>This option <code class="docutils literal notranslate"><span class="pre">output_html</span></code> will output all results into an html file (Results.html in the output_path folder) which is easy to view all pathways for each of the target compounds.  Note that this can only integrate figures from the <code class="docutils literal notranslate"><span class="pre">figures_graphviz</span></code> option.</p>
<p>Example of how to get Results.html file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">RetSynthGC</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">database_constraints</span><span class="o">=</span><span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span><span class="p">,</span> <span class="n">output_html</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figures_graphviz</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="ow">or</span>
<span class="n">rs</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">t</span> <span class="n">targetfile</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">db</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">db</span> <span class="o">-</span><span class="n">dbc</span> <span class="n">databasefile</span><span class="o">.</span><span class="n">constraints</span> <span class="o">--</span><span class="n">figures_graphviz</span> <span class="o">--</span><span class="n">output_html</span>
</pre></div>
</div>
</div>
<div class="section" id="other-options-in-retsynth">
<h2>5. Other options in RetSynth<a class="headerlink" href="#other-options-in-retsynth" title="Permalink to this headline">¶</a></h2>
<p>In RetSynth there are other options the user can specify to make their searches specific to their needs.</p>
<ol class="arabic simple">
<li><p>The option <code class="docutils literal notranslate"><span class="pre">--tan_thresh</span> <span class="pre">or</span> <span class="pre">--tanimoto_threshold</span></code> can be used if the inchi value is being used in the targetfile to specify the target compounds and if the <code class="docutils literal notranslate"><span class="pre">--inchidb</span></code> option is specified when generating a database(must be specifed when using option <code class="docutils literal notranslate"><span class="pre">-gdb</span></code> or when using an already developed database <code class="docutils literal notranslate"><span class="pre">-db</span></code>).  This option tells RetSynth to find pathways for all compounds in the database that have a structure similarity of the specified tanimoto score or greater (default 1) to the target compound in the target file.</p></li>
<li><p>The option <code class="docutils literal notranslate"><span class="pre">-k</span> <span class="pre">or</span> <span class="pre">--k_number_of_paths</span></code> specifies the number of shortest pathways (pathways with mimal number of reaction/enzyme pairs) wanted by the user. If the user specified <code class="docutils literal notranslate"><span class="pre">-k</span> <span class="pre">1</span></code> that would tell RetSynth to identify the shortest pathways and then next shortest pathways, giving the user pathways of 2 different lengths to the target compound where as <code class="docutils literal notranslate"><span class="pre">-k</span> <span class="pre">2</span></code> would result in pathways of three different lengths.  The default for -k is 0.</p></li>
<li><p>The options <code class="docutils literal notranslate"><span class="pre">-ms</span> <span class="pre">or</span> <span class="pre">--multiple_solutions</span></code> and <code class="docutils literal notranslate"><span class="pre">-cy</span> <span class="pre">or</span> <span class="pre">--cycles</span></code> indicate to RetSynth that all pathways having the minimal number of steps should be identified and that these pathways should not contain cycles.  Both of these option defaults are set true.</p></li>
</ol>
<div class="toctree-wrapper compound">
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="#"><span class="doc">Welcome to RetSynth’s documentation!</span></a></p></li>
<li><p><a class="reference internal" href="script.html"><span class="doc">Parameter Documentation</span></a></p></li>
<li><p><a class="reference internal" href="modules.html"><span class="doc">RetSynth Modules</span></a></p></li>
</ul>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="sphinxlocaltoc">
    <h3><a href="#">Page contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Welcome to RetSynth’s documentation!</a></li>
<li><a class="reference internal" href="#purpose">Purpose</a><ul>
<li><a class="reference internal" href="#installation">Installation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#retsynth-workflow">RetSynth Workflow</a><ul>
<li><a class="reference internal" href="#construction-of-metabolic-database">1. Construction of Metabolic Database</a></li>
<li><a class="reference internal" href="#generating-stoichiometric-matrix-for-reactions-and-compounds-in-the-metabolic-database">2. Generating Stoichiometric Matrix for Reactions and Compounds in the Metabolic Database</a></li>
<li><a class="reference internal" href="#target-compounds-input-file">3. Target Compounds (input file)</a></li>
<li><a class="reference internal" href="#examples-running-retsynth">4. Examples running RetSynth</a></li>
<li><a class="reference internal" href="#examples-for-running-retsynth">Examples for running RetSynth</a><ul>
<li><a class="reference internal" href="#results">Results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#gene-compatability-gc">Gene Compatability (GC)</a></li>
<li><a class="reference internal" href="#flux-balance-analysis-fba">Flux Balance Analysis (FBA)</a><ul>
<li><a class="reference internal" href="#fba-results">FBA Results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#visualization-chemdraw">Visualization_chemdraw</a></li>
<li><a class="reference internal" href="#visualization-graphviz">Visualization_graphviz</a></li>
<li><a class="reference internal" href="#visualize-results-in-html">Visualize results in html</a></li>
<li><a class="reference internal" href="#other-options-in-retsynth">5. Other options in RetSynth</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  </div>


  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
      <div class="relbar-bottom">
            
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
      <li><a href="#">RetSynth 1 documentation</a> &#187;</li>

        <li class="nav-item nav-item-this"><a href="">Welcome to RetSynth’s documentation!</a></li> 
      </ul>
    </div>
      </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Leanne Whitmore, Bernard Nguyen, Corey Hudson.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.1.2.
    </div>
      <!-- PSphinxTheme -->
  </body>
</html>